src/mono/mono/mini/interp/transform.c

```diff
@@ -2287,9 +2287,10 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
 
 			esize = mono_type_size (t, &align);
 			if (esize != 1) {
+				int power2 = mono_is_power_of_two ((guint32)esize);
 				g_assert (esize <= 32767);
-				interp_add_ins (td, MINT_MUL_P_IMM);
-				td->last_ins->data [0] = (gint16)esize;
+				interp_add_ins (td, power2 > 0 ? MINT_SHL_P_IMM : MINT_MUL_P_IMM);
+				td->last_ins->data [0] = (gint16)(power2 > 0 ? power2 : esize);
```

Comment:
Can't this just be a general optimization done against `MINT_MUL_P_IMM` rather than one done specifically here?

Reply:
I think this is one of those "we should always do this, even if we aren't optimizing" type transforms.

----------

That is, in RyuJIT we have different knobs that determine what type of optimizations can be done and there's some simple optimizations we do always regardless.

In particular we have a general `OptimizationDisabled` which returns true if we are in `Tier 0` or we are marked as needing to produce debuggable code (such as if a debugger is attached). This (and its logical inverse `OptimizationEnabled`) are used in most code to determine if we should optimize or not.

However, we also recognize that there's places where it is fundamentally beneficial to perform basic transforms regardless. For example, in `gtFoldExpr` rather than checking `OptimizationDisabled` we instead manually check if we need to produce debuggable code or if we are marked as `JIT_FLAG_MIN_OPT`, which is typically only set if the first compilation pass hit some internal limitation and failed to produce code (so it disables any and all optimizations as a fail safe so things still work, even if slow): https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/gentree.cpp#L13487-L13501 -- This is called unconditionally from places like import: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/importer.cpp#L7206

This same general consideration also applies to more general optimizations, such as in `Morph` (https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/morph.cpp#L13104), where we then skip additional constant folding if optimizations are disabled (https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/morph.cpp#L8288-L8292, unlike import which did this same bit unconditionally), but we do some other core optimizations such as replacing multiplication with shift regardless of optimization level: https://github.com/dotnet/runtime/blob/main/src/coreclr/jit/morph.cpp#L10456-L10520

This also allows code to do basic optimizations such as dead code elimination where trivially possible, leading to better throughput and handling, as we can differentiate between the levels of optimization and whether we need to be debuggable or not.

-------------

I imagine that Mono (JIT, LLVM, and Interpreter) would benefit from the same type of considerations and would see better perf and throughput if it mirrored the general premise

Reply:
> That would require some amount of constant folding or similar analysis though

Agreed. That's exactly why we do the basic constant folding during import in RyuJIT (only skipping them if we're in the true minopts mode or have a debugger attached) and then similarly do basic DCE based on this as well under similar circumstances.